src/data/list/basic.lean

```diff
@@ -1941,6 +1941,38 @@ begin
     apply hl _ (list.mem_cons_self _ _) }
 end
 
+/-- Transform a `l : list α` folded via `foldr` across a motive that holds
+for the seed element `b : β` and for all incremental `op : α → β → β`
+performed on the elements `(a : α) ∈ l`. -/
+def foldr_induction {C : β → Sort*} (l : list α) (op : α → β → β) (b : β)
+  (hl : ∀ (b : β) (hb : C b) (a : α) (ha : a ∈ l), C (op a b)) (hb : C b) :
+  C (list.foldr op b l) :=
+begin
+  induction l with hd tl IH,
+  { simpa using hb },
+  { rw list.foldr,
+    refine hl _ _ hd (mem_cons_self hd tl),
+    apply IH,
+    intros y hy x hx,
+    exact hl y hy x (mem_cons_of_mem hd hx) }
+end
+
+/-- Transform a `l : list α` folded via `foldl` across a motive that holds
+for the seed element `b : β` and for all incremental `op : β → α → β`
+performed on the elements `(a : α) ∈ l`. -/
+def foldl_induction {C : β → Sort*} (l : list α) (op : β → α → β) (b : β)
+  (hl : ∀ (b : β) (hb : C b) (a : α) (ha : a ∈ l), C (op b a)) (hb : C b) :
+  C (list.foldl op b l) :=
+begin
+  induction l with hd tl IH generalizing b,
+  { simpa using hb },
+  { rw list.foldl,
+    apply IH,
+    { intros y hy x hx,
+      exact hl y hy x (mem_cons_of_mem hd hx) },
+    { exact hl b hb hd (mem_cons_self hd tl) } }
+end
```

Comment:
This is slightly defintionally nicer as:
```
def foldl_induction {C : β → Sort*} (l : list α) (op : β → α → β) (b : β)
  (hb : C b) (hl : ∀ (b : β) (hb : C b) (a : α) (ha : a ∈ l), C (op b a)) :
  C (list.foldl op b l) :=
begin
  induction l with hd tl IH generalizing b,
  { refine cast _ hb, simp },
  { refine cast _ (IH _ (op b hd) _),
    { rw list.foldl, },
    { intros y hy x hx,
      exact hl y hy x (mem_cons_of_mem hd hx) },
    { exact hl b hb hd (mem_cons_self hd tl) } }
end
```
as then you end up with a single `cast` on each data path.

Probably more important is to add some simp lemmas for this recursor though.

Reply:
The easy one is:
```
lemma foldl_induction_nil {C : β → Sort*} (op : β → α → β) (b) (hb : C b) (hl) :
  foldl_induction op (nil : list α) b hb hl = hb := rfl
```